# Import all dependencies

In [1]:
import pandas as pd
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV

/home/frankjony17/Projects/venvs/ia-course/lib/python3.6/site-packages/pandas/compat/__init__.py:120: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


# Load datasets -> Dados e Clases

In [2]:
data = pd.read_csv("dataset/heart.csv")
x=data.iloc[:, 0:13]
y=data['target']

# Configures the model for training

In [3]:
def create_rede(optimizer, loss, kernel_initializer, activation, neurons):
    _classifier = Sequential()  # Sequential model.
    # Connected NN layer (Input Layer)
    _classifier.add(Dense(units=neurons, activation=activation, kernel_initializer=kernel_initializer, input_dim=13))
    # Hidden Layer
    _classifier.add(Dense(units=neurons, activation=activation, kernel_initializer=kernel_initializer))
    # Output Layer
    _classifier.add(Dense(units=1, activation='sigmoid'))
    # Configures the model for training
    _classifier.compile(optimizer=optimizer, loss=loss, metrics=['binary_accuracy'])
    return _classifier

# Implementation of the scikit-learn classifier API for Keras

In [4]:
classifier = KerasClassifier(build_fn=create_rede)

# Exhaustive search over specified parameter values for an estimator

Parameters:

In [5]:
parameters = {
    'batch_size': [5, 10],
    'epochs': [100, 200, 300],
    'optimizer': ['adam', 'sgd'],
    'loss': ['binary_crossentropy', 'hinge'],
    'kernel_initializer': ['random_uniform', 'normal'],
    'activation': ['relu', 'tanh'],
    'neurons': [4, 6, 8, 10]
}
# cv = cross-validation generator
grid_search = GridSearchCV(estimator=classifier, param_grid=parameters, scoring='accuracy', cv=5)

Run fit with all sets of parameters

In [ ]:
best_parameters = grid_search_fit.best_params_
better_accuracy = grid_search_fit.best_score_

In [ ]:
print(best_parameters)
print(better_accuracy)

{
    'activation': 'relu', 
    'batch_size': 20, 
    'epochs': 120, 
    'kernel_initializer': 'random_uniform',
    'loss': 'binary_crossentropy',
    'neurons': 16,
    'optimizer': 'adam'
}

0.9156652693681104